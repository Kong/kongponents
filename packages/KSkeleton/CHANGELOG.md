# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 5.0.9-alpha.0 (2021-11-16)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.8](https://github.com/Kong/kongponents/compare/v5.0.7...v5.0.8) (2021-11-15)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.7](https://github.com/Kong/kongponents/compare/v5.0.6...v5.0.7) (2021-10-26)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.6](https://github.com/Kong/kongponents/compare/v5.0.5...v5.0.6) (2021-10-26)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.5](https://github.com/Kong/kongponents/compare/v5.0.4...v5.0.5) (2021-10-13)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.4](https://github.com/Kong/kongponents/compare/v5.0.3...v5.0.4) (2021-10-06)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.3](https://github.com/Kong/kongponents/compare/v5.0.2...v5.0.3) (2021-09-22)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.2](https://github.com/Kong/kongponents/compare/v5.0.1...v5.0.2) (2021-09-15)

**Note:** Version bump only for package @kongponents/kskeleton





## [5.0.1](https://github.com/Kong/kongponents/compare/v5.0.0...v5.0.1) (2021-09-13)

**Note:** Version bump only for package @kongponents/kskeleton





# [5.0.0](https://github.com/Kong/kongponents/compare/v4.0.2...v5.0.0) (2021-09-13)

**Note:** Version bump only for package @kongponents/kskeleton





## [4.0.2](https://github.com/Kong/kongponents/compare/v4.0.1...v4.0.2) (2021-09-09)

**Note:** Version bump only for package @kongponents/kskeleton





## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/kskeleton





# 4.0.0 (2021-09-03)


### Bug Fixes

* **kskeleton:** button spinner ([8c32f9c](https://github.com/Kong/kongponents/commit/8c32f9c948d65517137fd4779200c0d58fe9a729))
* **skeleton:** correct the width classes ([#216](https://github.com/Kong/kongponents/issues/216)) ([7c08093](https://github.com/Kong/kongponents/commit/7c08093d510b932f3c6ec42e06de3456f9f190ae))


### Features

* **kemptystate:** make kemptystate support error state, add kskeleton  - intf-2312 ([#191](https://github.com/Kong/kongponents/issues/191)) ([180ff45](https://github.com/Kong/kongponents/commit/180ff45c225589f9736ae2285c81cd0cbd66ce30))
* **kskeleton:** add full screen loader ([#207](https://github.com/Kong/kongponents/issues/207)) ([e41f9d7](https://github.com/Kong/kongponents/commit/e41f9d77e090bb242cb0a1a37bb864a6562e6b60))
* **kskeleton:** switch from not rendering to transparency ([#281](https://github.com/Kong/kongponents/issues/281)) ([2d44e2d](https://github.com/Kong/kongponents/commit/2d44e2dc3d23c295e55b92570e3c9faedbd959bf))
* **skeleton:** add box primitive ([#213](https://github.com/Kong/kongponents/issues/213)) ([dd0681e](https://github.com/Kong/kongponents/commit/dd0681e715df5d92c36bb25304fd69cdcb0e7830))
* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [1.2.1](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@1.2.0...@kongponents/kskeleton@1.2.1) (2021-01-22)


### Bug Fixes

* **kskeleton:** button spinner ([8c32f9c](https://github.com/Kong/kongponents/commit/8c32f9c948d65517137fd4779200c0d58fe9a729))





# [1.2.0](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@1.0.2...@kongponents/kskeleton@1.2.0) (2020-12-18)


### Features

* **kskeleton:** switch from not rendering to transparency ([#281](https://github.com/Kong/kongponents/issues/281)) ([2d44e2d](https://github.com/Kong/kongponents/commit/2d44e2dc3d23c295e55b92570e3c9faedbd959bf))





## [1.0.2](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@1.0.0...@kongponents/kskeleton@1.0.2) (2020-06-25)

**Note:** Version bump only for package @kongponents/kskeleton





# [1.0.0](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.3.1...@kongponents/kskeleton@1.0.0) (2020-06-25)


### Features

* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [0.3.1](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.3.0...@kongponents/kskeleton@0.3.1) (2020-05-18)


### Bug Fixes

* **skeleton:** correct the width classes ([#216](https://github.com/Kong/kongponents/issues/216)) ([7c08093](https://github.com/Kong/kongponents/commit/7c08093d510b932f3c6ec42e06de3456f9f190ae))





# [0.3.0](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.2.1...@kongponents/kskeleton@0.3.0) (2020-05-18)


### Features

* **skeleton:** add box primitive ([#213](https://github.com/Kong/kongponents/issues/213)) ([dd0681e](https://github.com/Kong/kongponents/commit/dd0681e715df5d92c36bb25304fd69cdcb0e7830))





## [0.2.1](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.2.0...@kongponents/kskeleton@0.2.1) (2020-04-20)

**Note:** Version bump only for package @kongponents/kskeleton





# [0.2.0](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.1.5...@kongponents/kskeleton@0.2.0) (2020-04-20)


### Features

* **kskeleton:** add full screen loader ([#207](https://github.com/Kong/kongponents/issues/207)) ([e41f9d7](https://github.com/Kong/kongponents/commit/e41f9d77e090bb242cb0a1a37bb864a6562e6b60))





## [0.1.5](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.1.4...@kongponents/kskeleton@0.1.5) (2020-03-06)

**Note:** Version bump only for package @kongponents/kskeleton





## [0.1.4](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.1.3...@kongponents/kskeleton@0.1.4) (2020-03-06)

**Note:** Version bump only for package @kongponents/kskeleton





## [0.1.3](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.1.2...@kongponents/kskeleton@0.1.3) (2020-03-06)

**Note:** Version bump only for package @kongponents/kskeleton





## [0.1.2](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@0.1.1...@kongponents/kskeleton@0.1.2) (2020-03-05)

**Note:** Version bump only for package @kongponents/kskeleton





## [0.1.1](https://github.com/Kong/kongponents/compare/@kongponents/kskeleton@1.1.0...@kongponents/kskeleton@0.1.1) (2020-03-05)

**Note:** Version bump only for package @kongponents/kskeleton





# 1.1.0 (2020-02-28)


### Features

* **kemptystate:** make kemptystate support error state, add kskeleton  - intf-2312 ([#191](https://github.com/Kong/kongponents/issues/191)) ([180ff45](https://github.com/Kong/kongponents/commit/180ff45c225589f9736ae2285c81cd0cbd66ce30))
