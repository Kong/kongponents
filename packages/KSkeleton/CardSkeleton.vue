<template functional>
  <div class="skeleton-card-wrapper">
    <div
      v-for="c in props.cardCount"
      :key="c"
      class="skeleton-card-column">
      <div class="skeleton-card">
        <div class="skeleton-card-header">
          <div class="box width-2 height-2" />
          <div class="box width-10 height-2" />
        </div>
        <div class="skeleton-card-subheader">
          <div class="box width-6 height-1" />
          <div class="box width-6 height-1" />
        </div>
        <div class="skeleton-card-footer">
          <div class="box width-5 height-1" />
          <div class="box width-5 height-1" />
          <div class="box width-5 height-1" />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'CardSkeleton',
  props: {
    cardCount: {
      type: Number,
      default: 1
    }
  }
}
</script>

<style lang="scss" scoped>
$borderColor: #e6e6e6;
.skeleton-card-wrapper {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
}

.skeleton-card-column {
  width: 33.33%;
  padding: 0 1rem 0 0;
  margin-bottom: 1rem;
}

.skeleton-card {
  display: flex;
  flex-direction: column;
  height: 324px;
  padding: 1rem;
  border-radius: 3px;
  border: 1px solid $borderColor;
  .skeleton-card-header {
    width: 100%;
    padding-bottom: 1rem;
    margin-bottom: 1rem;
    border-bottom: 1px dotted $borderColor;
    > div:nth-of-type(2) {
      margin-left: 1rem;
    }
  }
  .skeleton-card-subheader {
    display: flex;
    justify-content: space-between;
  }
  .skeleton-card-footer {
    display: flex;
    align-self: flex-end;
    justify-content: space-between;
    width: 100%;
    margin-top: auto;
    padding-top: 1rem;
    border-top: 1px solid $borderColor;
  }
}
</style>
