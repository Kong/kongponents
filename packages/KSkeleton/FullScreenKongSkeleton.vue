<template>
  <div
    class="fullscreen-loading-container"
    data-testid="full-screen-loader">
    <div>
      <img src="./kong-loader.gif">
      <div class="progress">
        <div
          :style="{ width: `${progression}%` }"
          class="progress-bar"
          role="progressbar" />
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    progress: {
      type: Number,
      default: null
    }
  },
  data () {
    return {
      timer: null,
      progressInternal: 0
    }
  },

  computed: {
    progression () {
      return this.progress !== null ? this.progress : this.progressInternal
    }
  },

  destroyed () {
    clearInterval(this.timer)
  },

  mounted () {
    // If progress is user determined, don't fire interval to simulate progress
    if (this.progress) {
      return
    }

    this.timer = setInterval(() => {
      if (this.progressInternal >= 100) {
        clearInterval(this.timer)
        this.progressInternal = 100
      }

      this.progressInternal = Math.min(this.progressInternal + Math.ceil(Math.random(10) * 30), 100)
    }, 200)
  }
}
</script>

<style lang="scss" scoped>
.fullscreen-loading-container {
  position: fixed;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 10500;
  flex-direction: column;
  margin: var(--KSkeletonFullScreenMargin, 0);
  background: var(--white, color(white));
  .progress {
    margin-top: 1rem;
    max-width: 350px;
    background-color: var(--grey-92, color(grey-92));
    border-radius: 0.5rem;
    .progress-bar {
      border-radius: 0.5rem;
      height: 5px;
      background-color: var(--KSkeletonFullScreenProgressColor, var(--blue-500, color(blue-500)));
    }
  }
}
</style>
