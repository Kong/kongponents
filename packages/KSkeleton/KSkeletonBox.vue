<template>
  <div
    :class="{[`width-${width}`]: true, [`height-${height}`]: true }"
    class="box"
  />
</template>

<script>
export default {
  name: 'KSkeletonBox',
  props: {
    width: {
      type: String,
      default: '1',
      validator: function (val) {
        return ['1', '2', '5', '6', '10', '25', '50', '75', '100'].includes(val)
      }
    },
    height: {
      type: String,
      default: '1',
      validator: function (val) {
        return ['1', '2'].includes(val)
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.box {
  display: inline-flex;
  border-radius: 3px;
  background: linear-gradient(
      -70deg,
      #f2f2f2 0%,
      #f2f2f2 40%,
      #f7f7f7 50%,
      #f2f2f2 60%,
      #f2f2f2 100%
    )
    repeat;
  background-size: 400% 100%;
  animation: gradient 1s ease infinite;

  // Provided box widths
  &.width {
    &-1 {
      width: 1rem;
    }
    &-2 {
      width: 2rem;
    }
    &-5 {
      width: 5rem;
    }
    &-6 {
      width: 6rem;
    }
    &-10 {
      width: 25%;
    }
    &-25 {
      width: 25%;
    }
    &-50 {
      width: 50%;
    }
    &-75 {
      width: 80%;
    }
    &-100 {
      width: 100%;
    }
  }
  // Provided box heights
  &.height {
    &-1 {
      height: 1rem;
    }
    &-2 {
      height: 2rem;
    }
  }
}

@keyframes gradient {
  0% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0% 50%;
  }
}
</style>
