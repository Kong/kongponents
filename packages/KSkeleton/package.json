{
  "name": "@kongponents/kskeleton",
  "version": "6.9.3",
  "description": "Loading/Skeleton state component.",
  "main": "dist/KSkeleton.umd.min.js",
  "componentName": "KSkeleton",
  "source": "index.js",
  "files": [
    "dist",
    "*.vue",
    "index.js",
    "kong-loader.gif"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kicon": "^6.9.3",
    "@kongponents/styles": "^6.9.3"
  }
}
