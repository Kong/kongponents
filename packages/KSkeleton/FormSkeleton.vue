<template>
  <div class="skeleton-form-wrapper">
    <div class="skeleton-form-row">
      <KSkeletonBox
        width="10"
        height="1" />
      <KSkeletonBox
        width="100"
        height="2" />
    </div>
    <div class="skeleton-form-row">
      <KSkeletonBox
        width="10"
        height="1" />
      <KSkeletonBox
        width="100"
        height="2" />
    </div>
    <div class="skeleton-form-row">
      <KSkeletonBox
        width="10"
        height="1" />
      <KSkeletonBox
        width="100"
        height="2" />
    </div>
    <div class="skeleton-form-row">
      <KSkeletonBox
        width="10"
        height="1" />
      <KSkeletonBox
        width="100"
        height="2" />
    </div>
  </div>
</template>

<script>
import KSkeletonBox from './KSkeletonBox'

export default {
  name: 'FormSkeleton',
  components: { KSkeletonBox }
}
</script>

<style lang="scss" scoped>
.skeleton-form-wrapper {
  width: 100%;
  .skeleton-form-row {
    margin-bottom: 2rem;
  }
}
</style>
