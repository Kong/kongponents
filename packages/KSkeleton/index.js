import KSkeleton from './KSkeleton'
import KSkeletonBox from './KSkeletonBox'

const Plugin = {
  install (Vue) {
    Vue.component('KSkeleton', KSkeleton)
    Vue.component('KSkeletonBox', KSkeletonBox)
  }
}

export default Plugin
export { KSkeleton, KSkeletonBox }

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(Plugin)
}
