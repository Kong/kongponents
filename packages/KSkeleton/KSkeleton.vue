<template>
  <div
    :class="{ 'w-100': type !== 'spinner', 'opacity-0': !isVisible }"
    class="d-flex flex-wrap">
    <CardSkeleton
      v-if="type === 'card'"
      :card-count="cardCount ">
      <template v-slot:card-header>
        <slot name="card-header" />
      </template>
      <template v-slot:card-content>
        <slot name="card-content" />
      </template>
      <template v-slot:card-footer>
        <slot name="card-footer" />
      </template>
    </CardSkeleton>
    <TableSkeleton
      v-else-if="type === 'table'"
      :columns="tableColumns"
      :rows="tableRows">
      <slot />
    </TableSkeleton>
    <FormSkeleton v-else-if="type === 'form'"/>
    <FullScreenKongSkeleton
      v-else-if="type === 'fullscreen-kong'"
      :progress="progress"/>
    <KIcon
      v-else-if="type === 'spinner'"
      icon="spinner"
      view-box="0 0 16 16"
      color="#000"/>
    <Skeleton v-else />
  </div>
</template>

<script>
import Skeleton from './Skeleton'
import CardSkeleton from './CardSkeleton'
import TableSkeleton from './TableSkeleton'
import FormSkeleton from './FormSkeleton'
import FullScreenKongSkeleton from './FullScreenKongSkeleton'
import KIcon from '@kongponents/kicon/KIcon.vue'

export default {
  name: 'KSkeleton',
  components: {
    Skeleton,
    CardSkeleton,
    TableSkeleton,
    FormSkeleton,
    FullScreenKongSkeleton,
    KIcon
  },
  props: {
    delayMilliseconds: {
      type: Number,
      required: false,
      default: 750
    },
    type: {
      type: String,
      default: '',
      validator: (val) => [
        'table',
        'card',
        'form',
        'spinner',
        'fullscreen-kong',
        '' // default Skeleton.vue
      ].indexOf(val) !== -1
    },
    progress: {
      type: Number,
      required: false,
      default: null
    },
    cardCount: {
      type: Number,
      default: 1
    },
    tableColumns: {
      type: Number,
      required: false,
      default: 6
    },
    tableRows: {
      type: Number,
      required: false,
      default: 6
    }
  },
  data () {
    return {
      isVisible: false
    }
  },

  mounted () {
    setTimeout(() => {
      this.isVisible = true
    }, this.delayMilliseconds)
  }
}
</script>
<style>
.opacity-0 .box {
  opacity: 0;
}
.opacity-0 .kong-icon {
  display: none;
}
.opacity-0.mr-2 {
  margin-right: 0 !important;
}
</style>
