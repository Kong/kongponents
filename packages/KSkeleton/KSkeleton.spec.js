import { mount } from '@vue/test-utils'
import KSkeleton from '@/KSkeleton/KSkeleton'

describe('KSkeleton', () => {
  it('renders generic skeleton state by default', () => {
    const wrapper = mount(KSkeleton, {
      propsData: {
        delayMilliseconds: 0
      }
    })

    expect(wrapper.find('.skeleton-loader').exists()).toBe(true)
  })

  it('renders form skeleton state with 4 rows by default', () => {
    const wrapper = mount(KSkeleton, {
      propsData: {
        type: 'form',
        delayMilliseconds: 0
      }
    })

    expect(wrapper.find('.skeleton-form-wrapper').exists()).toBe(true)
    expect(wrapper.findAll('.skeleton-form-row')).toHaveLength(4)
  })

  it('renders card skeleton state with 2 cards', () => {
    const wrapper = mount(KSkeleton, {
      propsData: {
        type: 'card',
        cardCount: 2,
        delayMilliseconds: 0
      }
    })

    expect(wrapper.find('.skeleton-card-wrapper').exists()).toBe(true)
    expect(wrapper.findAll('.skeleton-card')).toHaveLength(2)
  })

  it('renders table skeleton state with 6 rows by default', () => {
    const wrapper = mount(KSkeleton, {
      propsData: {
        type: 'table',
        delayMilliseconds: 0
      }
    })

    expect(wrapper.find('.skeleton-table-wrapper').exists()).toBe(true)
    expect(wrapper.findAll('.skeleton-table-row')).toHaveLength(6)
  })

  it('renders spinner skeleton state', () => {
    const wrapper = mount(KSkeleton, {
      propsData: {
        type: 'spinner',
        delayMilliseconds: 0
      }
    })

    expect(wrapper.find('.kong-icon').exists()).toBe(true)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KSkeleton)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
