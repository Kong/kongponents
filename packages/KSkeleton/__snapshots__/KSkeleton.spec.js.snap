// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KSkeleton KSkeletonBox renders width and height 1`] = `<div class="box width-1 height-1"></div>`;

exports[`KSkeleton KSkeletonBox renders width and height 2`] = `<div class="box width-100 height-2"></div>`;

exports[`KSkeleton variants matches snapshot 1`] = `
<div class="d-flex flex-wrap w-100 opacity-0">
  <div class="skeleton-loader">
    <div class="box width-100 height-1"></div>
    <div class="box width-100 height-1"></div>
    <div class="box width-75 height-1"></div>
  </div>
</div>
`;
